### ISSUE TITLE
📎 Implement rule `useTopLevelRegex`

### ISSUE THREAD

**Issue by ematipico** [biomejs/biome#2148]
### Description

I *don't* know if there's a similar rule out there, we can add a reference later.

I came up with this rule while looking at many JS PRs out there, and I was surprised that there are many developers that don't adopt the following practice. Given the following code, the rule should trigger a case similar to this:

```js

function foo(someString) {
	return /[a-Z]*/.test(someString)
}
```

The rule should suggest moving the regex in a top-level variable:

```js

const REGEX = /[a-Z]*/;

function foo(someString) {
	return REGEX.test(someString)
}
```

The rule should not provide a code action; it should only suggest moving the regex in a variable at the top level of the module.

The reason for this rule is simple: regex requires a parsing phase, which could be expensive for some long ones. V8 and other engines have a long history of parsing, so I'm sure there is some optimisation, but recreating a new regex every time we call the function `foo` is definitely not optimal.

**Comment by arendjr** [comment]
I like the suggestion :+1:

Of course there is an obvious downside to it: Placing all regular expressions in the top-level scope will cause startup performance of the bundle to suffer. Usually this is negligible, and if the regexes are used inside a hot loop, it's absolutely the trade-off you want to make. But forcing all regexes to be top-level may swing performance too much in the other direction, because for code paths that are rarely taken, it does make sense to delay the compilation cost.

So I'm not sure I would recommend enabling the rule by default, even if it's a valuable rule to have in our arsenal :)

**Comment by RiESAEX** [comment]
Because Regex has internal state, lifting or inlining it is a danger task for both humans and compilers.
FYI: https://v8.dev/blog/regexp-tier-up

**Comment by dyc3** [comment]
I'd like to give this a shot.
